### ISSUE TITLE
Reducing binary size by breaking implicit dependency of spvtools_opt  on spvtools_val

### ISSUE THREAD

**Issue by amaiorano** [KhronosGroup/SPIRV-Tools#3759]
Hello, I'm on the Google Chrome GPU team, specifically on the [SwiftShader project](https://swiftshader.googlesource.com/SwiftShader), a Vulkan driver implementation that targets CPUs. I've been investigating reducing the size of our binaries, and SPIRV-Tools adds a significant amount.

For SwiftShader, we only require the SPIR-V optimization feature; however, linking against only `spvtools_opt` results in linker errors because of a couple of functions from the `spvtools_val` target:

```
lld-link: error: undefined symbol: spvValidateBinary
>>> referenced by .\..\..\third_party\SPIRV-Tools\src\source\libspirv.cpp:108
>>>               obj/third_party/SPIRV-Tools/src\spvtools/libspirv.obj:(public: bool __cdecl spvtools::SpirvTools::Validate(class std::__1::vector<unsigned int, class std::__1::allocator<unsigned int>> const &) const)
>>> referenced by .\..\..\third_party\SPIRV-Tools\src\source\libspirv.cpp:113
>>>               obj/third_party/SPIRV-Tools/src\spvtools/libspirv.obj:(public: bool __cdecl spvtools::SpirvTools::Validate(unsigned int const *, unsigned __int64) const)

lld-link: error: undefined symbol: spvValidateWithOptions
>>> referenced by .\..\..\third_party\SPIRV-Tools\src\source\libspirv.cpp:121
>>>               obj/third_party/SPIRV-Tools/src\spvtools/libspirv.obj:(public: bool __cdecl spvtools::SpirvTools::Validate(unsigned int const *, unsigned __int64, struct spv_validator_options_t *) const)
```

This results from a couple of calls being made to `SpirvTools::Validate()`, one in `Optimizer::Run()` and one in `PassManager::Run()`. When I stub the two missing functions, `spvValidateBinary` and `spvValidateWithOptions`, our binary size is reduced by about 500 KB.

For now, I'm going with stubbing these two functions; however, I think it would be useful if there was a way to link against only `spvtools_opt` without requiring `spvtools_val`. I'm not sure what the best approach would be here, though, as the decision to validate is based on runtime flags, and the binary size reduction would have to happen at compile/link time. Perhaps there can be overloads of `Optimizer::Run()` and `PassManager::Run()` that do not call `SpirvTools::Validate()`, which would rely on the linker removing the reference to these functions.

Alternately, perhaps a build-time flag could be exposed to stub out the implementations of the `SpirvTools::Validate()` overloads to do nothing and return success. However, this option would need to be exposed in the different build systems (GN in our case, for Chrome), and we'd need it set only for our non-debug build. This is trickier than deciding which function to call.

Note that there are other teams/projects at Google that would benefit greatly from this type of binary size reduction, as SPIRVTools is used by projects like [ANGLE](https://chromium.googlesource.com/angle/angle/) and [Dawn](https://dawn.googlesource.com/dawn).

Would love to hear any thoughts or ideas for this.

**Comment by alan-baker** [comment]
The optimizer assumes valid SPIR-V when it runs and may well crash or do other bad things on an invalid binary so I'm not sure that's the safest decision.

**Comment by amaiorano** [comment]
> The optimizer assumes valid SPIR-V when it runs and may well crash or do other bad things on an invalid binary so I'm not sure that's the safest decision.

Yes, but validation isn't enabled by default when optimizing, it's an option that you need to enable. Further, in our case, we know we have valid SPIR-V as it was validated beforehand.

**Comment by alan-baker** [comment]
If you validated the binary already, then you have SPIRV-Tools linked already? Is this a question of symbol visibility? Are you using static or shared libraries?

**Comment by amaiorano** [comment]
> If you validated the binary already, then you have SPIRV-Tools linked already? Is this a question of symbol visibility? Are you using static or shared libraries?

A few things:
1. It's actually the application's responsibility to send valid SPIR-V to the driver, not the driver's. It's undefined behaviour to send invalid SPIR-V, but not illegal.
2. In our case, ANGLE makes sure to produce valid SPIR-V, and sends it to the Vulkan driver.
3. We are statically linking to SPIRV-Tools as the fact that we use SPIRV-Tools (opt) is an implementation detail. It's not about symbol visibility, it's about making our SwiftShader DLL smaller for Chrome.

What I'm asking for here is a way to link against spirv-opt without spirv-val for our "official" Chrome build (the one released to the public). For internal builds, we will keep validation enabled to help us track down any bugs in SwiftShader or ANGLE, etc.

**Comment by dneto0** [comment]
Hey @amaiorano I'm also on the Chrome team, also working on Chrome's WebGPU flow.   I'll reach out to you internally.

**Comment by dj2** [comment]
Although the behaviour maybe undefined, you probably don't want that to be nullptr accesses or other crashing bugs which is what can happen in the optimizer if the SPIR-V isn't validated. It's only really safe to run the optimizer if you've run the validator ahead of time.
